File: motion_profile/v_maxa.py
import matplotlib.pyplot as plt
import numpy as np
import math

V0 = 0.0
Vn = 0.5
A0 = 1.0
An = 0.0001
MaxA = 1.124975

T11 = 3*((-V0 + Vn)*math.sqrt(9*A0**2 - A0*MaxA + A0*An - 9*A0 + MaxA**2 - MaxA*An) - (V0 - Vn)*(A0 + MaxA + An))/(-8*A0**2 + 3*A0*MaxA + A0*An + 9*A0 + 3*MaxA*An + An**2)
T22 = 3*(V0 - Vn)*(-A0 - MaxA - An + math.sqrt(9*A0**2 - A0*MaxA + A0*An - 9*A0 + MaxA**2 - MaxA*An))/(-8*A0**2 + 3*A0*MaxA + A0*An + 9*A0 + 3*MaxA*An + An**2)

print('T11 = ', T11)
print('T22 = ', T22)

T = T22#min([n for n in [T11, T22]  if n>0])

T2 = T*T
T3 = T2*T

A = np.array([[1, 0,   0,    0], 
              [0, 1,   0,    0], 
              [1, T,  T2,   T3], 
              [0, 1, 2*T, 3*T2]])
B = np.array([V0, A0, Vn, An])
#B = B.reshape([-1, 1])
a = np.linalg.inv(A).dot(B)


'''
m = np.array( [1, T, T*T, T*T*T, 0, 1, 2*T, 3*T*T])
det =  1./(m[2] * m[7] - m[3] * m[6])
c = [V0, 
     A0,
     det*( V0*( -m[0] * m[7] + m[4] * m[3] ) + A0*( -m[1] * m[7] + m[3] * m[5] ) + Vn*( m[7] ) + An*( -m[3] ) ),
     det*( V0*( m[0] * m[6] - m[4] * m[2] ) + A0*( m[1] * m[6] - m[2] * m[5] ) + Vn*( -m[6] ) + An*( m[2] ) )]
print(c)
'''


t1 = np.linspace(0, abs(T), abs(T*1000))
t2 = t1*t1
t3 = t2*t1

pos  = 0 + a[0]*t1 + a[1]*t2/2 + a[2]*t2/3 + a[3]*t3/4
vel  = a[0] + a[1]*t1 + a[2]*t2 + a[3]*t3
acc  = a[1] + 2*a[2]*t1 + 3*a[3]*t2
jerk = 2*a[2] + 6*a[3]*t1

for i in range(4):
  print('a[', i ,'] = ', a[i])

plt.suptitle('v maxa')
plt.subplot(411)
plt.xlabel('t')
plt.ylabel('pos')
plt.plot(t1, pos)
plt.subplot(412)
plt.xlabel('t')
plt.ylabel('vel')
plt.plot(t1, vel)
plt.subplot(413)
plt.xlabel('t')
plt.ylabel('acc')
plt.plot(t1, acc)
plt.subplot(414)
plt.ylabel('jerk')
plt.plot(t1, jerk)
plt.show()

File: motion_profile/v_maxa.c
#include <math.h>
#include <stdio.h>

double vt = 0.0;
double vn = 0.5;
double at = 0.0;
double an = 0.0;
double amax = 1.12;

/*bool MatInvert(float m[16], float invOut[16]) {
	float inv[16], det;
	int i;
 
	inv[ 0] =  m[5] * m[10] * m[15] - m[5] * m[11] * m[14] - m[9] * m[6] * m[15] + m[9] * m[7] * m[14] + m[13] * m[6] * m[11] - m[13] * m[7] * m[10];
	inv[ 4] = -m[4] * m[10] * m[15] + m[4] * m[11] * m[14] + m[8] * m[6] * m[15] - m[8] * m[7] * m[14] - m[12] * m[6] * m[11] + m[12] * m[7] * m[10];
	inv[ 8] =  m[4] * m[ 9] * m[15] - m[4] * m[11] * m[13] - m[8] * m[5] * m[15] + m[8] * m[7] * m[13] + m[12] * m[5] * m[11] - m[12] * m[7] * m[ 9];
	inv[12] = -m[4] * m[ 9] * m[14] + m[4] * m[10] * m[13] + m[8] * m[5] * m[14] - m[8] * m[6] * m[13] - m[12] * m[5] * m[10] + m[12] * m[6] * m[ 9];
	inv[ 1] = -m[1] * m[10] * m[15] + m[1] * m[11] * m[14] + m[9] * m[2] * m[15] - m[9] * m[3] * m[14] - m[13] * m[2] * m[11] + m[13] * m[3] * m[10];
	inv[ 5] =  m[0] * m[10] * m[15] - m[0] * m[11] * m[14] - m[8] * m[2] * m[15] + m[8] * m[3] * m[14] + m[12] * m[2] * m[11] - m[12] * m[3] * m[10];
	inv[ 9] = -m[0] * m[ 9] * m[15] + m[0] * m[11] * m[13] + m[8] * m[1] * m[15] - m[8] * m[3] * m[13] - m[12] * m[1] * m[11] + m[12] * m[3] * m[ 9];
	inv[13] =  m[0] * m[ 9] * m[14] - m[0] * m[10] * m[13] - m[8] * m[1] * m[14] + m[8] * m[2] * m[13] + m[12] * m[1] * m[10] - m[12] * m[2] * m[ 9];
	inv[ 2] =  m[1] * m[ 6] * m[15] - m[1] * m[ 7] * m[14] - m[5] * m[2] * m[15] + m[5] * m[3] * m[14] + m[13] * m[2] * m[ 7] - m[13] * m[3] * m[ 6];
	inv[ 6] = -m[0] * m[ 6] * m[15] + m[0] * m[ 7] * m[14] + m[4] * m[2] * m[15] - m[4] * m[3] * m[14] - m[12] * m[2] * m[ 7] + m[12] * m[3] * m[ 6];
	inv[10] =  m[0] * m[ 5] * m[15] - m[0] * m[ 7] * m[13] - m[4] * m[1] * m[15] + m[4] * m[3] * m[13] + m[12] * m[1] * m[ 7] - m[12] * m[3] * m[ 5];
	inv[14] = -m[0] * m[ 5] * m[14] + m[0] * m[ 6] * m[13] + m[4] * m[1] * m[14] - m[4] * m[2] * m[13] - m[12] * m[1] * m[ 6] + m[12] * m[2] * m[ 5];
	inv[ 3] = -m[1] * m[ 6] * m[11] + m[1] * m[ 7] * m[10] + m[5] * m[2] * m[11] - m[5] * m[3] * m[10] - m[ 9] * m[2] * m[ 7] + m[ 9] * m[3] * m[ 6];
	inv[ 7] =  m[0] * m[ 6] * m[11] - m[0] * m[ 7] * m[10] - m[4] * m[2] * m[11] + m[4] * m[3] * m[10] + m[ 8] * m[2] * m[ 7] - m[ 8] * m[3] * m[ 6];
	inv[11] = -m[0] * m[ 5] * m[11] + m[0] * m[ 7] * m[ 9] + m[4] * m[1] * m[11] - m[4] * m[3] * m[ 9] - m[ 8] * m[1] * m[ 7] + m[ 8] * m[3] * m[ 5];
	inv[15] =  m[0] * m[ 5] * m[10] - m[0] * m[ 6] * m[ 9] - m[4] * m[1] * m[10] + m[4] * m[2] * m[ 9] + m[ 8] * m[1] * m[ 6] - m[ 8] * m[2] * m[ 5];
 
	det = m[0] * inv[0] + m[1] * inv[4] + m[2] * inv[8] + m[3] * inv[12];
 
	if(det == 0)
		return false;
 
	det = 1.f / det;
 
	for(i = 0; i < 16; i++)
		invOut[i] = inv[i] * det;
 
	return true;
}*/
  
int main(int argc, char** argv) {
	double T22=0.0, T11=0.0;
	if (at == 0.0 && an == 0.0) {
		T22 = abs( 3*(vn - v0)/(2*amax) );
	} else {
		T11 = 3*((-vt + vn)*sqrt(9*at*at - at*amax + at*an - 9*at + amax*amax - amax*an) - (vt - at)*(at + amax + an))/(-8*at*at + 3*at*amax + at*an + 9*at + 3*amax*an + an*an);
		T22 = 3*(vt - vn)*(-at - amax - an + sqrt(9*at*at - at*amax + at*an - 9*at + amax*amax - amax*an))/(-8*at*at + 3*at*amax + at*an + 9*at + 3*amax*an + an*an);
	}
	
	double T = T22;
	double m[8] = {1, T, T*T, T*T*T, 0, 1, 2*T, 3*T*T};
	
	printf("T11 = %f\n", T11);
	printf("T22 = %f\n", T22);
	
	double a[4];
	double det =  1./(m[2] * m[7] - m[3] * m[6]);
	
	a[0] = vt;
	a[1] = at;
	a[2] = det*( vt*( -m[0] * m[7] + m[4] * m[3] ) + at*( -m[1] * m[7] + m[3] * m[5] ) + vn*( m[7] ) + an*( -m[3] ) );
	a[3] = det*( vt*( m[0] * m[6] - m[4] * m[2] ) + at*( m[1] * m[6] - m[2] * m[5] ) + vn*( -m[6] ) + an*( m[2] ) );
	
	
	
}
